README
id3v2 - A command line id3v2 tag editor.  
http://id3v2.sourceforge.net

In case you didn't know id3 tags can be found on mp3 files, they can store
infomation about what band recorded the song, the song name, etc.

I wrote this because id3v1 are seriously deficient in what kind of info can
be in them and the length of info that can be in them.  id3v2 tags are much
better but there isn't a tool to edit them in Linux.  This is a tool to fill
that space.

See INSTALL for directions on building id3v2.

As of version 0.1.10 this code is under the LGPL.  A lot of it is based off
the examples in id3lib, but there is some code from me :).

		- myers <myers_carpenter at users.sourceforge.net>

thanks goes out to:

Antti Andreimann - patch to write only one version of tag

Vladimir Pastukhov - patch to handle General object frames w/o segfaulting

Errica Crome - patch to fix genre display.

Branko Radovanovic <branko-r at users.sf.net> - 
  patch 1:
    - No more seg faults for: --IPLS, --UFID, --PCNT, --
    POPM. Note that these still merely print the info (i.e. the
    file is *not* changed).
    
    - No more seg faults for --comment. And the new code
    actually *sets* the comment.
    
    - No more seg faults for --TXXX desc:text.
    
    - "TSIZ results in unrecognized option" now fixed.
    
    Also rearranged some redundant code and made several
    minor changes (a bit more consistent coding style,
    spelling, etc.) 

  patch 2:
     
    In PrintID3v1Tag() fread() in fact returns size_t (number
    of items read), which is unsigned, and the error
    reporting code is thus never executed. Fixed now.
    
    BTW, it would probably be a good idea to compile with "-
    W -Wall pedantic". This is how I caught it.    
  (I've changed the make file as he suggested)

Jim Ursetto <ursetto at users.sf.net> -
    Attached patch deletes id3v2 tags when passed an empty
    string, rather than just setting them to a null string.
    I.e. -a "" will delete the artist tag. Works for
    comments too: -c "ID3v1 Comment::XXX" will delete the
    default comment derived from the v1 tag.

Martin Soto <soto at informatik.uni-kl.de>
	When a file has an unknown genre (number 255) the program dies with a
	segmentation fault.  When applied to list.cpp the patch should correct
	that by printing "Unknown" as the genre.

Aubin Paul <aubin at punknews.org>
	patch to build under gcc 3.2
	
Jens Rehsack <rehsack at liwing.de>
    I attach a few patches for the id3v2 tool you're responsible for. Maybe
    it's useful to apply them, 'cause at the moment the implementation
    doesn't conforms the ANSI C++ standard. (I didn't check the whole source
    but those lines which causes the error).

Warren Stevens <warren at catzooks.com>
	Whole bunch of stuff

Roberto Costa <roberto.costa at ensta.org>	
	Patch to fix id3 genres
	

debian/changelog
id3v2 (0.1.7-1) unstable; urgency=low

  * Fixed display/setting genre.

 -- Myers Carpenter <myers@fil.org>  Mon, 15 Apr 2002 21:34:40 -0400

id3v2 (0.1.6-1) unstable; urgency=low

  * Fixed perms on updated file.

 -- Myers Carpenter <myers@fil.org>  Tue, 19 Mar 2002 10:15:03 -0500

id3v2 (0.1.3-1) unstable; urgency=low

  * Initial Release.

 -- Robert Woodcock <rcw@debian.org>  Thu, 18 May 2000 21:29:11 -0700

Local variables:
mode: debian-changelog
End:


genre.h
/* id3 Renamer
 * genre.h - Header for music genres
 * Copyright (C) 1998  Robert Alto (badcrc@tscnet.com)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
 * USA.
 */

#ifndef __GENRE_H__
#define __GENRE_H__

extern int GetGenreCount();
extern const char *GetGenreFromNum(int);
extern int GetNumFromGenre(const char*);

#endif /* __GENRE_H__ */


list.cpp
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//  
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//  
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

#if defined HAVE_CONFIG_H
#include <config.h>
#endif

#include <iostream>
#include <cstring>
#include <id3/tag.h>
#include <getopt.h>
#include <cstdlib>
#include <cstdio>
#include <id3/misc_support.h>
#include "frametable.h"
#include "genre.h"

char *GetDescription(const ID3_FrameID eFrameID)
{
  for(int ii = 0; ii < frameTableCount; ii++ )
  {
    if (frameTable[ii].frameID == eFrameID)
      return frameTable[ii].frameLongName;
  }
  return NULL;
}

void PrintFrameHelp(char *sName)
{
    for(int ii = 0; ii < frameTableCount - 1; ii++ )
    {
      std::cout << "    --" << frameTable[ii].frameName << "    " 
           << frameTable[ii].frameLongName << std::endl;
    }
    return;
}

void PrintGenreList()
{
  for (int ii = 0; ii < GetGenreCount(); ii++)
    printf("%3d: %s\n", ii, GetGenreFromNum(ii));
}

int PrintInformation(char *sFileName, const ID3_Tag &myTag)
{
  bool firstLine = true;
  const ID3_Frame * myFrame;
  ID3_Tag::ConstIterator *Iter=myTag.CreateIterator();
  for (size_t nFrames = 0; nFrames < myTag.NumFrames(); nFrames++)
  {
    myFrame = Iter->GetNext();

    if(firstLine) {
      std::cout << "id3v2 tag info for " << sFileName << ":" << std::endl;
      firstLine = false;
    }

    if (NULL != myFrame)
    { 
      const char* desc = myFrame->GetDescription();
      if (!desc) desc = "";
      std::cout << myFrame->GetTextID() << " (" << desc << "): ";
      ID3_FrameID eFrameID = myFrame->GetID();

      switch (eFrameID)
      {
        case ID3FID_ALBUM:
        case ID3FID_BPM:
        case ID3FID_COMPOSER:
        case ID3FID_COPYRIGHT:
        case ID3FID_DATE:
        case ID3FID_PLAYLISTDELAY:
        case ID3FID_ENCODEDBY:
        case ID3FID_LYRICIST:
        case ID3FID_FILETYPE:
        case ID3FID_TIME:
        case ID3FID_CONTENTGROUP:
        case ID3FID_TITLE:
        case ID3FID_SUBTITLE:
        case ID3FID_INITIALKEY:
        case ID3FID_LANGUAGE:
        case ID3FID_SONGLEN:
        case ID3FID_MEDIATYPE:
        case ID3FID_ORIGALBUM:
        case ID3FID_ORIGFILENAME:
        case ID3FID_ORIGLYRICIST:
        case ID3FID_ORIGARTIST:
        case ID3FID_ORIGYEAR:
        case ID3FID_FILEOWNER:
        case ID3FID_LEADARTIST:
        case ID3FID_BAND:
        case ID3FID_CONDUCTOR:
        case ID3FID_MIXARTIST:
        case ID3FID_PARTINSET:
        case ID3FID_PUBLISHER:
        case ID3FID_TRACKNUM:
        case ID3FID_RECORDINGDATES:
        case ID3FID_NETRADIOSTATION:
        case ID3FID_NETRADIOOWNER:
        case ID3FID_SIZE:
        case ID3FID_ISRC:
        case ID3FID_ENCODERSETTINGS:
        case ID3FID_YEAR:
        {
          char *sText = ID3_GetString(myFrame, ID3FN_TEXT);
          std::cout << sText << std::endl;
          delete [] sText;
          break;
        }
        case ID3FID_CONTENTTYPE:
        {
          const char* genre_str;
          int genre_id = 255;
          char *sText = ID3_GetString(myFrame, ID3FN_TEXT);
          sscanf(sText, "(%d)", &genre_id);
          if (genre_id == 255) {
            genre_str = sText;
            genre_id = GetNumFromGenre(sText);
          }
          else
            genre_str = GetGenreFromNum(genre_id);
          std::cout << genre_str << " (" << genre_id << ")" << std::endl;
          delete [] sText;
          break;
        }
        case ID3FID_USERTEXT:
        {
          char 
            *sText = ID3_GetString(myFrame, ID3FN_TEXT), 
            *sDesc = ID3_GetString(myFrame, ID3FN_DESCRIPTION);
          std::cout << "(" << sDesc << "): " << sText << std::endl;
          delete [] sText;
          delete [] sDesc;
          break;
        }
        case ID3FID_COMMENT:
        case ID3FID_UNSYNCEDLYRICS:
        {
          char 
            *sText = ID3_GetString(myFrame, ID3FN_TEXT), 
            *sDesc = ID3_GetString(myFrame, ID3FN_DESCRIPTION), 
            *sLang = ID3_GetString(myFrame, ID3FN_LANGUAGE);
          std::cout << "(" << sDesc << ")[" << sLang << "]: "
               << sText << std::endl;
          delete [] sText;
          delete [] sDesc;
          delete [] sLang;
          break;
        }
        case ID3FID_WWWAUDIOFILE:
        case ID3FID_WWWARTIST:
        case ID3FID_WWWAUDIOSOURCE:
        case ID3FID_WWWCOMMERCIALINFO:
        case ID3FID_WWWCOPYRIGHT:
        case ID3FID_WWWPUBLISHER:
        case ID3FID_WWWPAYMENT:
        case ID3FID_WWWRADIOPAGE:
        {
          char *sURL = ID3_GetString(myFrame, ID3FN_URL);
          std::cout << sURL << std::endl;
          delete [] sURL;
          break;
        }
        case ID3FID_WWWUSER:
        {
          char 
            *sURL = ID3_GetString(myFrame, ID3FN_URL),
            *sDesc = ID3_GetString(myFrame, ID3FN_DESCRIPTION);
          std::cout << "(" << sDesc << "): " << sURL << std::endl;
          delete [] sURL;
          delete [] sDesc;
          break;
        }
        case ID3FID_INVOLVEDPEOPLE:
        {
          // This isn't the right way to do it---will only get first person
          size_t nItems = myFrame->Field(ID3FN_TEXT).GetNumTextItems();
          for (size_t nIndex = 1; nIndex <= nItems; nIndex++)
          {
            char *sPeople = ID3_GetString(myFrame, ID3FN_TEXT, nIndex);
            std::cout << sPeople;
            delete [] sPeople;
            if (nIndex < nItems)
            {
              std::cout << ", ";
            }
          }
          std::cout << std::endl;
          break;
        }
        case ID3FID_PICTURE:
        {
          char
            *sMimeType = ID3_GetString(myFrame, ID3FN_MIMETYPE),
            *sDesc     = ID3_GetString(myFrame, ID3FN_DESCRIPTION),
            *sFormat   = ID3_GetString(myFrame, ID3FN_IMAGEFORMAT);
          size_t
            nPicType   = myFrame->Field(ID3FN_PICTURETYPE).Get(),
            nDataSize  = myFrame->Field(ID3FN_DATA).Size();
          std::cout << "(" << sDesc << ")[" << sFormat << ", "
               << nPicType << "]: " << sMimeType << ", " << nDataSize
               << " bytes" << std::endl;
          delete [] sMimeType;
          delete [] sDesc;
          delete [] sFormat;
          break;
        }
        case ID3FID_GENERALOBJECT:
        {
          char 
          *sMimeType = ID3_GetString(myFrame, ID3FN_MIMETYPE), 
          *sDesc = ID3_GetString(myFrame, ID3FN_DESCRIPTION), 
          *sFileName = ID3_GetString(myFrame, ID3FN_FILENAME);
          size_t nDataSize = myFrame->GetField(ID3FN_DATA)->Size();
          std::cout << "(" << sDesc << ")[" 
               << sFileName << "]: " << sMimeType << ", " << nDataSize
               << " bytes" << std::endl;
          delete [] sMimeType;
          delete [] sDesc;
          delete [] sFileName;
          break;
        }
        case ID3FID_UNIQUEFILEID:
        {
          char *sOwner = ID3_GetString(myFrame, ID3FN_OWNER);
          size_t nDataSize = myFrame->Field(ID3FN_DATA).Size();
          std::cout << sOwner << ", " << nDataSize
               << " bytes" << std::endl;
          delete [] sOwner;
          break;
        }
        case ID3FID_PLAYCOUNTER:
        {
          size_t nCounter = myFrame->Field(ID3FN_COUNTER).Get();
          std::cout << nCounter << std::endl;
          break;
        }
        case ID3FID_POPULARIMETER:
        {
          char *sEmail = ID3_GetString(myFrame, ID3FN_EMAIL);
          size_t
            nCounter = myFrame->Field(ID3FN_COUNTER).Get(),
            nRating = myFrame->Field(ID3FN_RATING).Get();
          std::cout << sEmail << ", counter=" 
               << nCounter << " rating=" << nRating;
          delete [] sEmail;
          break;
        }
        case ID3FID_CRYPTOREG:
        case ID3FID_GROUPINGREG:
        {
          char *sOwner = ID3_GetString(myFrame, ID3FN_OWNER);
          size_t 
            nSymbol = myFrame->Field(ID3FN_ID).Get(),
            nDataSize = myFrame->Field(ID3FN_DATA).Size();
          std::cout << "(" << nSymbol << "): " << sOwner
               << ", " << nDataSize << " bytes";
          break;
        }
        case ID3FID_AUDIOCRYPTO:
        case ID3FID_EQUALIZATION:
        case ID3FID_EVENTTIMING:
        case ID3FID_CDID:
        case ID3FID_MPEGLOOKUP:
        case ID3FID_OWNERSHIP:
        case ID3FID_PRIVATE:
        case ID3FID_POSITIONSYNC:
        case ID3FID_BUFFERSIZE:
        case ID3FID_VOLUMEADJ:
        case ID3FID_REVERB:
        case ID3FID_SYNCEDLYRICS:
        case ID3FID_SYNCEDTEMPO:
        case ID3FID_METACRYPTO:
        {
          std::cout << " (unimplemented)" << std::endl;
          break;
        }
        default:
        {
          std::cout << " frame" << std::endl;
          break;
        }
      }
    }
  }
  if(firstLine)
    return 1;
  
  return 0;
}

int PrintID3v1Tag(char *sFileName) 
// code from id3 
{
  struct id3 {
    char tag[3];
    char title[30];
    char artist[30];
    char album[30];
    char year[4];
    /* With ID3 v1.0, the comment is 30 chars long */
    /* With ID3 v1.1, if comment[28] == 0 then comment[29] == tracknum */
    char comment[30];
    unsigned char genre;
  } id3v1tag;
  FILE *fp;
  
  fp = fopen(sFileName, "r"); /* read only */

  if (fp == NULL) { /* file didn't open */
    fprintf(stderr, "fopen: %s: ", sFileName);
    perror("id3v2");
    return -1;
  }
  if (fseek(fp, -128, SEEK_END) < 0) {
    /* problem rewinding */
  } else { /* we rewound successfully */ 
    if (fread(&id3v1tag, 128, 1, fp) != 1) {
      /* read error */
      fprintf(stderr, "fread: %s: ", sFileName);
      perror("");
    }
  }
    
  fclose(fp);

    /* This simple detection code has a 1 in 16777216
     * chance of misrecognizing or deleting the last 128
     * bytes of your mp3 if it isn't tagged. ID3 ain't
     * world peace, live with it.
     */
    
  
  if (!strncmp(id3v1tag.tag, "TAG", 3))
  {
    std::cout << "id3v1 tag info for " << sFileName << ":" << std::endl;
    printf("Title  : %-30.30s  Artist: %-30.30s\n",
            id3v1tag.title, id3v1tag.artist);
    printf("Album  : %-30.30s  Year: %-4.4s, Genre: %s (%d)\n",
            id3v1tag.album, id3v1tag.year, 
            (id3v1tag.genre < GetGenreCount())
            ? GetGenreFromNum(id3v1tag.genre) : 
            "Unknown", id3v1tag.genre);
    if (!id3v1tag.comment[28])
      printf("Comment: %-28.28s    Track: %d\n", 
             id3v1tag.comment, id3v1tag.comment[29]);
    else
      printf("Comment: %-30.30s\n", id3v1tag.comment);
  } 
  else 
  {
    return 1;     
  }
  return 0;
}

void ListTag(int argc, char *argv[], int optind, int rfc822)
{
  bool tags = false;
  int ret;
  for (size_t nIndex = optind; nIndex < argc; nIndex++)
  {
    ID3_Tag myTag;

    ret = PrintID3v1Tag(argv[nIndex]);
    if (ret == -1)
    {
      continue;
    } 
    else if(ret == 0)
    {
      tags = true;
    }
    myTag.Link(argv[nIndex], ID3TT_ID3V2);
    if(!PrintInformation(argv[nIndex],myTag))
      tags = true;
    if(!tags)
      std::cout << argv[nIndex] << ": No ID3 tag" << std::endl;
  }

  return;
}



convert.cpp
#include <cstdio>
#include <iostream>
#include <id3/tag.h>
#include <cstdlib>

void DeleteTag(int argc, char *argv[], int optind, int whichTags)
{

    FILE * fp;
    
    for (size_t nIndex = optind; nIndex < argc; nIndex++)
    {
      /* cludgy to check if we have the proper perms */
      fp = fopen(argv[nIndex], "r+");
      if (fp == NULL) { /* file didn't open */
        fprintf(stderr, "fopen: %s: ", argv[nIndex]);
        perror("id3v2");
        continue;
      }
      fclose(fp);

      ID3_Tag myTag;

      std::cout << "Stripping id3 tag in \"";
      std::cout << argv[nIndex] << "\"...";

      myTag.Clear();
      myTag.Link(argv[nIndex], ID3TT_ALL);

      luint nTags;
      switch(whichTags) 
      {
        case 1: 
          nTags = myTag.Strip(ID3TT_ID3V1);
          std::cout << "id3v1 ";
          break;
        case 2:
          nTags = myTag.Strip(ID3TT_ID3V2);
          std::cout << "id3v2 ";
          break; 
        case 0:
        default:
          nTags = myTag.Strip(ID3TT_ID3);
          std::cout << "id3v1 and v2 ";
      }

      std::cout << "stripped." << std::endl;
    }

  return;
}


void ConvertTag(int argc, char *argv[], int optind)
{
    for (size_t nIndex = optind; nIndex < argc; nIndex++)
    {
      ID3_Tag myTag;

      std::cout << "Converting id3v1 tag to id3v2 in ";
      std::cout << argv[nIndex] << "...";

      myTag.Clear();
      myTag.Link(argv[nIndex], ID3TT_ALL);

      luint nTags;

      nTags = myTag.Update(ID3TT_ID3V2);
      std::cout << " converted ";
      std::cout << std::endl;
    }

  return;
}


frametable.h
/* id3v2 frame table
 * frametable.h - struct with info on all id3v2 frames
 * Copyright (C) 2000 Myers Carpenter (icepick@sourceforge.net)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
 * USA.
 */

#ifndef __FRAMETABLE_H__
#define __FRAMETABLE_H__

#include <id3/globals.h>

struct frameTbl {
  char *frameName;
  enum ID3_FrameID frameID;
  char *frameLongName;
}; 

static struct frameTbl frameTable[] = {
  { "AENC", ID3FID_AUDIOCRYPTO,  	    "Audio encryption" },
  { "APIC", ID3FID_PICTURE,	 			    "Attached picture" },
  { "COMM", ID3FID_COMMENT,	 			    "Comments" },
  { "COMR", ID3FID_COMMERCIAL,	 	    "Commercial frame" },
  { "ENCR", ID3FID_CRYPTOREG,	 		    "Encryption method registration" },
  { "EQUA", ID3FID_EQUALIZATION, 	    "Equalization" },
	{ "ETCO", ID3FID_EVENTTIMING, 	    "Event timing codes" },
	{ "GEOB", ID3FID_GENERALOBJECT,	    "General encapsulated object" },
	{ "GRID", ID3FID_GROUPINGREG,       "Group identification registration" },
	{ "IPLS", ID3FID_INVOLVEDPEOPLE,    "Involved people list" },
	{ "LINK", ID3FID_LINKEDINFO,        "Linked information" },
	{ "MCDI", ID3FID_CDID,              "Music CD identifier" },
	{ "MLLT", ID3FID_MPEGLOOKUP,        "MPEG location lookup table" },
	{ "OWNE", ID3FID_OWNERSHIP,         "Ownership frame" },
	{ "PRIV", ID3FID_PRIVATE,           "Private frame" },
	{ "PCNT", ID3FID_PLAYCOUNTER,       "Play counter" },
	{ "POPM", ID3FID_POPULARIMETER,     "Popularimeter" },
	{ "POSS", ID3FID_POSITIONSYNC,      "Position synchronisation frame" },
	{ "RBUF", ID3FID_BUFFERSIZE,        "Recommended buffer size" },
	{ "RVAD", ID3FID_VOLUMEADJ,         "Relative volume adjustment" },
	{ "RVRB", ID3FID_REVERB,            "Reverb" },
	{ "SYLT", ID3FID_SYNCEDLYRICS,      "Synchronized lyric/text" },
	{ "SYTC", ID3FID_SYNCEDTEMPO,       "Synchronized tempo codes" },
	{ "TALB", ID3FID_ALBUM,             "Album/Movie/Show title" },
	{ "TBPM", ID3FID_BPM,               "BPM (beats per minute)" },
	{ "TCOM", ID3FID_COMPOSER,          "Composer" },
	{ "TCON", ID3FID_CONTENTTYPE,       "Content type" },
	{ "TCOP", ID3FID_COPYRIGHT,         "Copyright message" },
	{ "TDAT", ID3FID_DATE,              "Date" },
	{ "TDLY", ID3FID_PLAYLISTDELAY,     "Playlist delay" },
	{ "TENC", ID3FID_ENCODEDBY,         "Encoded by" },
	{ "TEXT", ID3FID_LYRICIST,          "Lyricist/Text writer" },
	{ "TFLT", ID3FID_FILETYPE,          "File type" },
	{ "TIME", ID3FID_TIME,              "Time" },
	{ "TIT1", ID3FID_CONTENTGROUP,      "Content group description" },
	{ "TIT2", ID3FID_TITLE,             "Title/songname/content description" },
	{ "TIT3", ID3FID_SUBTITLE,          "Subtitle/Description refinement" },
	{ "TKEY", ID3FID_INITIALKEY,        "Initial key" },
	{ "TLAN", ID3FID_LANGUAGE,          "Language(s)" },
	{ "TLEN", ID3FID_SONGLEN,           "Length" },
	{ "TMED", ID3FID_MEDIATYPE,         "Media type" },
	{ "TOAL", ID3FID_ORIGALBUM,         "Original album/movie/show title" },
	{ "TOFN", ID3FID_ORIGFILENAME,      "Original filename" },
	{ "TOLY", ID3FID_ORIGLYRICIST,      "Original lyricist(s)/text writer(s)" },
	{ "TOPE", ID3FID_ORIGARTIST,        "Original artist(s)/performer(s)" },
	{ "TORY", ID3FID_ORIGYEAR,          "Original release year" },
	{ "TOWN", ID3FID_FILEOWNER,         "File owner/licensee" },
	{ "TPE1", ID3FID_LEADARTIST,        "Lead performer(s)/Soloist(s)" },
	{ "TPE2", ID3FID_BAND,              "Band/orchestra/accompaniment" },
	{ "TPE3", ID3FID_CONDUCTOR,         "Conductor/performer refinement" },
	{ "TPE4", ID3FID_MIXARTIST,         "Interpreted, remixed, or otherwise modified by" },
	{ "TPOS", ID3FID_PARTINSET,         "Part of a set" },
	{ "TPUB", ID3FID_PUBLISHER,         "Publisher" },
	{ "TRCK", ID3FID_TRACKNUM,          "Track number/Position in set" },
	{ "TRDA", ID3FID_RECORDINGDATES,    "Recording dates" }, 
	{ "TRSN", ID3FID_NETRADIOSTATION,   "Internet radio station name" },
	{ "TRSO", ID3FID_NETRADIOOWNER,     "Internet radio station owner" },
	{ "TSIZ", ID3FID_SIZE,              "Size" },
	{ "TSRC", ID3FID_ISRC,              "ISRC (international standard recording code)" },
	{ "TSSE", ID3FID_ENCODERSETTINGS,   "Software/Hardware and settings used for encoding" },
  { "TXXX", ID3FID_USERTEXT,          "User defined text information" },
  { "TYER", ID3FID_YEAR,              "Year" },
  { "UFID", ID3FID_UNIQUEFILEID,      "Unique file identifier" },
  { "USER", ID3FID_TERMSOFUSE,        "Terms of use" },
  { "USLT", ID3FID_UNSYNCEDLYRICS,    "Unsynchronized lyric/text transcription" },
  { "WCOM", ID3FID_WWWCOMMERCIALINFO, "Commercial information" },
  { "WCOP", ID3FID_WWWCOPYRIGHT,      "Copyright/Legal infromation" },
  { "WOAF", ID3FID_WWWAUDIOFILE,      "Official audio file webpage" },
  { "WOAR", ID3FID_WWWARTIST,         "Official artist/performer webpage" },
  { "WOAS", ID3FID_WWWAUDIOSOURCE,    "Official audio source webpage" },
  { "WORS", ID3FID_WWWRADIOPAGE,      "Official internet radio station homepage" },
  { "WPAY", ID3FID_WWWPAYMENT,        "Payment" },
  { "WPUB", ID3FID_WWWPUBLISHER,      "Official publisher webpage" },
  { "WXXX", ID3FID_WWWUSER,           "User defined URL link" },
  { "    ", ID3FID_METACRYPTO,        "Encrypted meta frame" },
  { "????", ID3FID_NOFRAME,           "Error" }
};

int frameTableCount = 75;


#endif /* __FRAMETABLE_H__ */


create_map.cpp
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//  
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//  
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


#include <iostream>
#include <stdlib.h>
#include "genre.cpp"

#ifndef SORT_RUNTIME
int CompareGenres(const void *a, const void *b)
{
    int g1, g2;
    g1 = *((int*)a);
    g2 = *((int*)b);

    return strcmp(GetGenreFromNum(g1), GetGenreFromNum(g2));
}
#endif  // SORT_RUNTIME

void main()
{
    int i;
    for (i=0; i<GetGenreCount(); ++i)
        genre_map[i] = i;

    qsort(genre_map, GetGenreCount(), sizeof(int), &CompareGenres);

    cout << "char *genre_table[] = {" << genre_map[0];
    for (i=1; i<GetGenreCount(); ++i)
        cout << ", " << genre_map[i];
    cout << "};" << std::endl;
}


genre.cpp
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//  
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//  
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

#if defined HAVE_CONFIG_H
#include <config.h>
#endif

#ifndef __GENRE_H__
#include "genre.h"
#endif

#ifdef SORT_RUNTIME
#include <stdlib.h>
#endif  // SORT_RUNTIME
#include <string.h>


char *genre_table[] = {"Blues",
                       "Classic Rock",
                       "Country",
                       "Dance",
                       "Disco",
                       "Funk",
                       "Grunge",
                       "Hip-Hop",
                       "Jazz",
                       "Metal",
                       "New Age",
                       "Oldies",
                       "Other",
                       "Pop",
                       "R&B",
                       "Rap",
                       "Reggae",
                       "Rock",
                       "Techno",
                       "Industrial",
                       "Alternative",
                       "Ska",
                       "Death Metal",
                       "Pranks",
                       "Soundtrack",
                       "Euro-Techno",
                       "Ambient",
                       "Trip-Hop",
                       "Vocal",
                       "Jazz+Funk",
                       "Fusion",
                       "Trance",
                       "Classical",
                       "Instrumental",
                       "Acid",
                       "House",
                       "Game",
                       "Sound Clip",
                       "Gospel",
                       "Noise",
                       "Alt. Rock",
                       "Bass",
                       "Soul",
                       "Punk",
                       "Space",
                       "Meditative",
                       "Instrum. Pop",
                       "Instrum. Rock",
                       "Ethnic",
                       "Gothic",
                       "Darkwave",
                       "Techno-Indust.",
                       "Electronic",
                       "Pop-Folk",
                       "Eurodance",
                       "Dream",
                       "Southern Rock",
                       "Comedy",
                       "Cult",
                       "Gangsta",
                       "Top 40",
                       "Christian Rap",
                       "Pop/Funk",
                       "Jungle",
                       "Native American",
                       "Cabaret",
                       "New Wave",
                       "Psychadelic",
                       "Rave",
                       "Showtunes",
                       "Trailer",
                       "Lo-Fi",
                       "Tribal",
                       "Acid Punk",
                       "Acid Jazz",
                       "Polka",
                       "Retro",
                       "Musical",
                       "Rock & Roll",
                       "Hard Rock",
                       "Folk",
                       "Folk/Rock",
                       "National Folk",
                       "Swing",
                       "Fusion",
                       "Bebob",
                       "Latin",
                       "Revival",
                       "Celtic",
                       "Bluegrass",
                       "Avantgarde",
                       "Gothic Rock",
                       "Progress. Rock",
                       "Psychadel. Rock",
                       "Symphonic Rock",
                       "Slow Rock",
                       "Big Band",
                       "Chorus",
                       "Easy Listening",
                       "Acoustic",
                       "Humour",
                       "Speech",
                       "Chanson",
                       "Opera",
                       "Chamber Music",
                       "Sonata",
                       "Symphony",
                       "Booty Bass",
                       "Primus",
                       "Porn Groove",
                       "Satire",
 /* The following were added 1999 26 Apr by Ben Gertzfield <che@debian.org> 
  * as per the list at http://mp3.musichall.cz/id3master/faq.htm but with a 
  * few spell-checks confirmed by running 'strings' on the in_mp3.dll file 
  * from winamp 2.10 (sorry :)
  */
		       "Slow Jam",
		       "Club",
		       "Tango",
		       "Samba",
		       "Folklore",
		       "Ballad",
		       "Power Ballad",
		       "Rhythmic Soul",
		       "Freestyle",
		       "Duet",
		       "Punk Rock",
		       "Drum Solo",
		       "A Capella",
		       "Euro-House",
		       "Dance Hall",
		       "Goa",
		       "Drum & Bass",
		       "Club-House",
		       "Hardcore",
		       "Terror",
		       "Indie",
		       "BritPop",
		       "Negerpunk",
		       "Polsk Punk",
		       "Beat",
		       "Christian Gangsta Rap",	/* DJ JC in da house */
		       "Heavy Metal",
		       "Black Metal",
		       "Crossover",
		       "Contemporary Christian",
		       "Christian Rock",
				/* winamp 1.91 genres */
		       "Merengue",
		       "Salsa",
		       "Thrash Metal",
				/* winamp 1.92 genres */
		       "Anime",
		       "Jpop",
		       "Synthpop",
                       };

const int genre_count = sizeof(genre_table) / sizeof(char*);

#ifdef SORT_RUNTIME
int genre_map[genre_count];
#else
int genre_map[genre_count] =
    { 123, 34, 74, 73, 99, 40, 20, 26, 145, 90, 116, 41, 135, 85, 96, 138, 89,
      0, 107, 132, 65, 88, 104, 102, 97, 136, 61, 141, 1, 32, 112, 128, 57,
      140, 2, 139, 58, 3, 125, 50, 22, 4, 55, 127, 122, 120, 98, 52, 48, 124,
      25, 54, 80, 81, 115, 119, 5, 30, 84, 36, 59, 126, 38, 49, 91, 6, 79, 129,
      137, 7, 35, 100, 131, 19, 46, 47, 33, 8, 29, 146, 63, 86, 71, 45, 142, 9,
      77, 82, 64, 133, 10, 66, 39, 11, 103, 12, 75, 134, 13, 53, 62, 109, 117,
      23, 108, 92, 93, 67, 43, 121, 14, 15, 68, 16, 76, 87, 118, 17, 78, 143,
      114, 110, 69, 21, 111, 95, 105, 42, 37, 24, 56, 44, 101, 83, 94, 106,
      147, 113, 18, 51, 130, 144, 60, 70, 31, 72, 27, 28 } ;
#endif  // SORT_RUNTIME

int GetGenreCount()
{
    return genre_count;
}

const char *GetGenreFromNum(int genre_id)
{
    if ((genre_id >= 0) && (genre_id < GetGenreCount()))
        return genre_table[genre_id];
    else
        return "Unknown";
}

int GetNumFromGenre(const char *genre)
{
    int a=0;
    int b=GetGenreCount()-1;
    int i;

    while(a <= b) {
        const int i = (a+b)/2;
        const int genre_id = genre_map[i];
        const int cmp = strcmp(GetGenreFromNum(genre_id), genre);

        if (cmp == 0)
            return genre_id;
        else if (cmp < 0)
            a = i+1;
        else
            b = i-1;
    }

    return 255;
}

#ifdef SORT_RUNTIME
int CompareGenres(const void *a, const void *b)
{
    int g1, g2;
    g1 = *((int*)a);
    g2 = *((int*)b);

    return strcmp(GetGenreFromNum(g1), GetGenreFromNum(g2));
}

void InitGenres()
{
    int i;
    for (i=0; i<GetGenreCount(); ++i)
        genre_map[i] = i;

    qsort(genre_map, GetGenreCount(), sizeof(int), &CompareGenres);
}
#endif  // SORT_RUNTIME
